text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/tag/builtin/BuiltInTags.java

```diff
@@ -0,0 +1,186 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.text.minimessage.tag.builtin;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Stream;
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.kyori.adventure.text.format.TextDecoration;
+import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver;
+
+/**
+ * Tag types distributed with MiniMessage.
+ *
+ * <p>All built-in types are included in the default tag resolver.</p>
+ *
+ * @since 4.10.0
+ */
+public final class BuiltInTags {
+  private BuiltInTags() {
+  }
+
+  private static final TagResolver DECORATION = Stream.concat(TextDecoration.NAMES.keyToValue().entrySet().stream(), DecorationTag.DECORATION_ALIASES.entrySet().stream())
+    .flatMap(entry -> {
+      return Stream.of(
+        TagResolver.resolver(entry.getKey(), (args, ctx) -> DecorationTag.create(entry.getValue(), args, ctx)),
+        TagResolver.resolver(DecorationTag.REVERT + entry.getKey(), DecorationTag.createNegated(entry.getValue()))
+        );
+    })
+    .collect(TagResolver.toTagResolver());
+  private static final TagResolver COLOR = new ColorTagResolver();
+  private static final TagResolver HOVER_EVENT = TagResolver.resolver(HoverTag.HOVER, HoverTag::create);
+  private static final TagResolver CLICK_EVENT = TagResolver.resolver(ClickTag.CLICK, ClickTag::create);
+  private static final TagResolver KEYBIND = TagResolver.resolver(KeybindTag.KEYBIND, KeybindTag::create);
+  private static final TagResolver TRANSLATABLE = TagResolver.resolver(
+    names(TranslatableTag.TRANSLATABLE, TranslatableTag.TRANSLATABLE_2, TranslatableTag.TRANSLATABLE_3),
+    TranslatableTag::create
+  );
+  private static final TagResolver INSERTION = TagResolver.resolver(InsertionTag.INSERTION, InsertionTag::create);
+  private static final TagResolver FONT = TagResolver.resolver(FontTag.FONT, FontTag::create);
+  private static final TagResolver GRADIENT = TagResolver.resolver(GradientTag.GRADIENT, GradientTag::create);
+  private static final TagResolver RAINBOW = TagResolver.resolver(RainbowTag.RAINBOW, RainbowTag::create);
+
+  /**
+   * Get a resolver for the {@value HoverTag#HOVER} tag.
+   *
+   * @return a resolver for the {@value HoverTag#HOVER} tag
+   * @since 4.10.0
+   */
+  public static TagResolver hoverEvent() {
```

Comment:
is there a reason to use methods over just sticking with the constants?

Reply:
Better encapsulation, and lets us change the layout of tags down the line and provide the same API surface for tags that are provided with MiniMessage but not part of the standard collection/created by default. Specifically I'm thinking of the PR to add a color tag that supports CSS colour names -- we could have `BuiltInTags.color()` for the default flag, that just returns a field, and `BuiltInTags.color(Map<String, TextColor> aliases)` that creates a new `TagResolver` each time. Users wouldn't have to think about both field-tags and method-tags.